{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Bacterial Ensemble Language Model\n",
    "\n",
    "This notebook trains a language model on the ensemble of bacterial genomes assembled in the [Bacterial Ensemble 0 Data Processing](https://github.com/kheyer/Genomic-ULMFiT/blob/master/Bacteria/Bacterial%20Ensemble/Bacterial%20Ensemble%200%20Data%20Processing.ipynb) notebook. The language model trained is based on the AWD-LSTM architecture. the genomic input information is split into 5-mers with a stride of 2 bases between each 5-mer. The model is trained to take an input sequence of 5-mers and predict the next 5-mer. This allows us to train a model that learns the structure of genomic information in a totally unsupervised way.\n",
    "\n",
    "The base of the language model (token embedding + LSTM layers) will then be used to initialize a classification model.\n",
    "\n",
    "For more detail on how genomic data is processed and how these language models are trained, see the following notebooks:\n",
    "\n",
    "[E. coli 1 Naive Model](https://github.com/kheyer/Genomic-ULMFiT/blob/master/Bacteria/E.%20Coli/E.%20coli%201%20Naive%20Model.ipynb)\n",
    "\n",
    "[E. coli 2 Genomic Pretraining](https://github.com/kheyer/Genomic-ULMFiT/blob/master/Bacteria/E.%20Coli/E.%20coli%202%20Genomic%20Pretraining.ipynb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%reload_ext autoreload\n",
    "%autoreload 2\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "# sets device for model and PyTorch tensors\n",
    "os.environ[\"CUDA_VISIBLE_DEVICES\"]=\"0\"\n",
    "#os.environ[\"CUDA_VISIBLE_DEVICES\"]=\"0,1,2,3,4,5,6,7\" # \"-1\" if want to run on a CPU, otherwise define the GPU number"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from fastai import *\n",
    "from fastai.text import *\n",
    "from Bio import Seq\n",
    "from Bio.Seq import Seq\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.SeqFeature import FeatureLocation, CompoundLocation\n",
    "import networkx as nx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "sys.path.append(\"../../..\")\n",
    "from utils import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "path = Path('/mnt/wd_4tb/shared_disk_wd4tb/mattscicluna/data/genomic_ulmfit/bacterial_genomes/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(path/'bacterial_data.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CATCGACGAGGGAACGCAGATTTTTCGTTAAAGCTCTGGCCCCTCG...</td>\n",
       "      <td>Corynebacterium striatum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TGGCTGGGTCTCGGCGATATAGAGGAAGTTGCCCTCACGCCCACGA...</td>\n",
       "      <td>Corynebacterium striatum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GTCGGTGCCGACGAAAGCGCGGGCAATCGTCATGGATTCATCAGCC...</td>\n",
       "      <td>Corynebacterium striatum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GAGTTTCTGGCTCAAGATTTCACGCACCGTGGCTTGATACACCATG...</td>\n",
       "      <td>Corynebacterium striatum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACGCTGAGATTCTTTCCGAGCCCATTTTCTAGCCTGTCCCATTGGC...</td>\n",
       "      <td>Corynebacterium striatum</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            Sequence                    Source\n",
       "0  CATCGACGAGGGAACGCAGATTTTTCGTTAAAGCTCTGGCCCCTCG...  Corynebacterium striatum\n",
       "1  TGGCTGGGTCTCGGCGATATAGAGGAAGTTGCCCTCACGCCCACGA...  Corynebacterium striatum\n",
       "2  GTCGGTGCCGACGAAAGCGCGGGCAATCGTCATGGATTCATCAGCC...  Corynebacterium striatum\n",
       "3  GAGTTTCTGGCTCAAGATTTCACGCACCGTGGCTTGATACACCATG...  Corynebacterium striatum\n",
       "4  ACGCTGAGATTCTTTCCGAGCCCATTTTCTAGCCTGTCCCATTGGC...  Corynebacterium striatum"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df[:100]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 10% of the data used for validation\n",
    "train_df, valid_df = split_data(df, 0.9)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((91, 2), (9, 2))"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_df.shape, valid_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "tok = Tokenizer(GenomicTokenizer, n_cpus=1, pre_rules=[], post_rules=[], special_cases=['xxpad'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = GenomicTextLMDataBunch.from_df(path, train_df, valid_df, bs=800, tokenizer=tok, text_cols=0, label_cols=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1025"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Model vocabulary - 1025 tokens. 1024 5-mer nucleotide combinations plus one padding token\n",
    "len(data.vocab.itos)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save model vocabulary - this will be important later\n",
    "np.save(path/'bact_vocab.npy', data.vocab.itos)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "voc = np.load(path/'bact_vocab.npy')\n",
    "model_vocab = GenomicVocab(voc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = GenomicTextLMDataBunch.from_df(path, train_df, valid_df, bs=800, tokenizer=tok, vocab=model_vocab,\n",
    "                              chunksize=80000, text_cols=0, label_cols=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1025, 1025)"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(data.vocab.itos), len(data.vocab.stoi)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "config = dict(emb_sz=400, n_hid=1150, n_layers=3, pad_token=0, qrnn=False, output_p=0.25, \n",
    "                          hidden_p=0.1, input_p=0.2, embed_p=0.02, weight_p=0.15, tie_weights=True, out_bias=True)\n",
    "drop_mult = 0.25"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "learn = get_model_LM(data, drop_mult, config)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SequentialRNN(\n",
       "  (0): AWD_LSTM(\n",
       "    (encoder): Embedding(1025, 400, padding_idx=0)\n",
       "    (encoder_dp): EmbeddingDropout(\n",
       "      (emb): Embedding(1025, 400, padding_idx=0)\n",
       "    )\n",
       "    (rnns): ModuleList(\n",
       "      (0): WeightDropout(\n",
       "        (module): LSTM(400, 1150, batch_first=True)\n",
       "      )\n",
       "      (1): WeightDropout(\n",
       "        (module): LSTM(1150, 1150, batch_first=True)\n",
       "      )\n",
       "      (2): WeightDropout(\n",
       "        (module): LSTM(1150, 400, batch_first=True)\n",
       "      )\n",
       "    )\n",
       "    (input_dp): RNNDropout()\n",
       "    (hidden_dps): ModuleList(\n",
       "      (0): RNNDropout()\n",
       "      (1): RNNDropout()\n",
       "      (2): RNNDropout()\n",
       "    )\n",
       "  )\n",
       "  (1): LinearDecoder(\n",
       "    (decoder): Linear(in_features=400, out_features=1025, bias=True)\n",
       "    (output_dp): RNNDropout()\n",
       "  )\n",
       ")"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "learn.model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "    <div>\n",
       "        <style>\n",
       "            /* Turns off some styling */\n",
       "            progress {\n",
       "                /* gets rid of default border in Firefox and Opera. */\n",
       "                border: none;\n",
       "                /* Needs to be in here for Safari polyfill so background images work as expected. */\n",
       "                background-size: auto;\n",
       "            }\n",
       "            .progress-bar-interrupted, .progress-bar-interrupted::-webkit-progress-bar {\n",
       "                background: #F44336;\n",
       "            }\n",
       "        </style>\n",
       "      <progress value='3' class='' max='4' style='width:300px; height:20px; vertical-align: middle;'></progress>\n",
       "      75.00% [3/4 00:09<00:03]\n",
       "    </div>\n",
       "    \n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: left;\">\n",
       "      <th>epoch</th>\n",
       "      <th>train_loss</th>\n",
       "      <th>valid_loss</th>\n",
       "      <th>accuracy</th>\n",
       "      <th>time</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>6.933172</td>\n",
       "      <td>#na#</td>\n",
       "      <td>00:03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>6.933215</td>\n",
       "      <td>#na#</td>\n",
       "      <td>00:03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>6.933174</td>\n",
       "      <td>#na#</td>\n",
       "      <td>00:03</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table><p>\n",
       "\n",
       "    <div>\n",
       "        <style>\n",
       "            /* Turns off some styling */\n",
       "            progress {\n",
       "                /* gets rid of default border in Firefox and Opera. */\n",
       "                border: none;\n",
       "                /* Needs to be in here for Safari polyfill so background images work as expected. */\n",
       "                background-size: auto;\n",
       "            }\n",
       "            .progress-bar-interrupted, .progress-bar-interrupted::-webkit-progress-bar {\n",
       "                background: #F44336;\n",
       "            }\n",
       "        </style>\n",
       "      <progress value='0' class='' max='1' style='width:300px; height:20px; vertical-align: middle;'></progress>\n",
       "      0.00% [0/1 00:00<00:00]\n",
       "    </div>\n",
       "    "
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LR Finder is complete, type {learner_name}.recorder.plot() to see the graph.\n"
     ]
    }
   ],
   "source": [
    "learn.lr_find(num_it=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZIAAAEGCAYAAABPdROvAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+WH4yJAAARH0lEQVR4nO3dfZBddX3H8fenCalVFLBEjQmYgFGbcSrgFqnY1qrYhFIS+whFjZQxw1iqjPUBqx2dtjPVaYsOloqpojC1om11SGsqRaoFa7FsFLBpQFLUEokQfOChMNK03/5xT+xlubu57G93b5Z9v2bu7D2/h/P7Xhbuh3POvWdTVUiSNF0/NOoCJEnzm0EiSWpikEiSmhgkkqQmBokkqcniURcwlw4//PBauXLlqMuQpHll27Ztd1XV0sn6F1SQrFy5kvHx8VGXIUnzSpJvTNXvqS1JUhODRJLUxCCRJDUxSCRJTQwSSVITg0SS1MQgkSQ1MUgkSU0MEklSE4NEktTEIJEkNTFIJElNDBJJUhODRJLUxCCRJDUxSCRJTQwSSVITg0SS1MQgkSQ1MUgkSU0MEklSE4NEktTEIJEkNTFIJElNDBJJUpORBkmStUluTrIzyXkD+pPkgq7/xiTHTehflOTLSf5u7qqWJPUbWZAkWQRcCKwD1gCnJ1kzYdg6YHX32AS8b0L/64Ads1yqJGkKozwiOR7YWVW3VtWDwGXA+glj1gOXVs+1wKFJlgEkWQH8PPCBuSxakvRQowyS5cBtfdu7urZhx7wHeBPwv1MtkmRTkvEk43v27GmrWJL0MKMMkgxoq2HGJDkFuLOqtu1vkaraXFVjVTW2dOnS6dQpSZrCKINkF3BE3/YK4PYhx5wInJrk6/ROib0oyV/MXqmSpMmMMkiuA1YnWZVkCXAasGXCmC3AK7tPb50A3F1Vu6vqLVW1oqpWdvP+sapePqfVS5IAWDyqhatqb5JzgCuARcDFVbU9ydld/0XAVuBkYCdwP3DmqOqVJA2WqomXJR69xsbGanx8fNRlSNK8kmRbVY1N1u832yVJTQwSSVITg0SS1MQgkSQ1MUgkSU0MEklSE4NEktTEIJEkNTFIJElNDBJJUhODRJLUxCCRJDUxSCRJTQwSSVITg0SS1MQgkSQ1MUgkSU0MEklSE4NEktTEIJEkNTFIJElNDBJJUhODRJLUxCCRJDUxSCRJTQwSSVITg0SS1MQgkSQ1MUgkSU0MEklSk5EGSZK1SW5OsjPJeQP6k+SCrv/GJMd17Uck+WySHUm2J3nd3FcvSYIRBkmSRcCFwDpgDXB6kjUThq0DVnePTcD7uva9wG9X1Y8BJwC/OWCuJGkOjPKI5HhgZ1XdWlUPApcB6yeMWQ9cWj3XAocmWVZVu6vqSwBVdS+wA1g+l8VLknpGGSTLgdv6tnfx8DDY75gkK4FjgS/OeIWSpP0aZZBkQFs9kjFJDgb+Bji3qu4ZuEiyKcl4kvE9e/ZMu1hJ0mCjDJJdwBF92yuA24cdk+QgeiHykar6xGSLVNXmqhqrqrGlS5fOSOGSpP83yiC5DlidZFWSJcBpwJYJY7YAr+w+vXUCcHdV7U4S4IPAjqo6f27LliT1Wzyqhatqb5JzgCuARcDFVbU9ydld/0XAVuBkYCdwP3BmN/1E4BXAV5Jc37X9TlVtncvXIEmCVE28LPHoNTY2VuPj46MuQ5LmlSTbqmpssn6/2S5JamKQSJKaGCSSpCYGiSSpiUEiSWpikEiSmhgkkqQmBokkqYlBIklqYpBIkpoYJJKkJgaJJKmJQSJJamKQSJKaGCSSpCYGiSSpiUEiSWpikEiSmhgkkqQmBokkqYlBIklqYpBIkpoYJJKkJgaJJKmJQSJJamKQSJKaDBUkSR6X5Ie6589IcmqSg2a3NEnSfDDsEcnVwGOSLAeuAs4EPjxbRUmS5o9hgyRVdT/wi8B7q+plwJrZK0uSNF8MHSRJfhI4A/hU17Z4dkqSJM0nwwbJucBbgE9W1fYkRwGfnb2yJEnzxVBBUlX/VFWnVtW7uovud1XVa1sXT7I2yc1JdiY5b0B/klzQ9d+Y5Lhh50qS5sawn9r6yyRPSPI44N+Bm5O8sWXhJIuAC4F19K63nJ5k4nWXdcDq7rEJeN8jmCtJmgPDntpaU1X3ABuArcCRwCsa1z4e2FlVt1bVg8BlwPoJY9YDl1bPtcChSZYNOVeSNAeGDZKDuu+NbAAur6r/Bqpx7eXAbX3bu7q2YcYMMxeAJJuSjCcZ37NnT2PJkqSJhg2S9wNfBx4HXJ3kacA9jWtnQNvEcJpszDBze41Vm6tqrKrGli5d+ghLlCTtz1Af4a2qC4AL+pq+keRnG9feBRzRt70CuH3IMUuGmCtJmgPDXmw/JMn5+04RJfkTekcnLa4DVidZlWQJcBqwZcKYLcAru09vnQDcXVW7h5wrSZoDw57auhi4F/jV7nEP8KGWhatqL3AOcAWwA/h49x2Vs5Oc3Q3bCtwK7AT+HHjNVHNb6pEkTU+q9n/NPMn1VXXM/toOdGNjYzU+Pj7qMiRpXkmyrarGJusf9ojkgSQv6NvpicADrcVJkua/Ye+XdTZwaZJDuu3vAhtnpyRJ0nwy7Ke2bgCek+QJ3fY9Sc4FbpzN4iRJB75H9BcSq+qe7hvuAK+fhXokSfNMy5/aHfSlQEnSAtMSJK23SJEkPQpMeY0kyb0MDowAPzIrFUmS5pUpg6SqHj9XhUiS5qeWU1uSJBkkkqQ2BokkqYlBIklqYpBIkpoYJJKkJgaJJKmJQSJJamKQSJKaGCSSpCYGiSSpiUEiSWpikEiSmhgkkqQmBokkqYlBIklqYpBIkpoYJJKkJgaJJKmJQSJJamKQSJKaGCSSpCYjCZIkT0xyZZJbup+HTTJubZKbk+xMcl5f+x8luSnJjUk+meTQuatektRvVEck5wFXVdVq4Kpu+yGSLAIuBNYBa4DTk6zpuq8Enl1VPw58FXjLnFQtSXqYUQXJeuCS7vklwIYBY44HdlbVrVX1IHBZN4+q+oeq2tuNuxZYMcv1SpImMaogeXJV7Qbofj5pwJjlwG1927u6tol+A/j7Ga9QkjSUxbO14ySfAZ4yoOutw+5iQFtNWOOtwF7gI1PUsQnYBHDkkUcOubQkaVizFiRV9ZLJ+pLckWRZVe1Osgy4c8CwXcARfdsrgNv79rEROAV4cVUVk6iqzcBmgLGxsUnHSZKmZ1SntrYAG7vnG4HLB4y5DlidZFWSJcBp3TySrAXeDJxaVffPQb2SpEmMKkjeCZyU5BbgpG6bJE9NshWgu5h+DnAFsAP4eFVt7+b/KfB44Mok1ye5aK5fgCSpZ9ZObU2lqr4NvHhA++3AyX3bW4GtA8Y9fVYLlCQNzW+2S5KaGCSSpCYGiSSpiUEiSWpikEiSmhgkkqQmBokkqYlBIklqYpBIkpoYJJKkJgaJJKmJQSJJamKQSJKaGCSSpCYGiSSpiUEiSWpikEiSmhgkkqQmBokkqYlBIklqYpBIkpoYJJKkJgaJJKmJQSJJamKQSJKaGCSSpCYGiSSpiUEiSWpikEiSmhgkkqQmBokkqclIgiTJE5NcmeSW7udhk4xbm+TmJDuTnDeg/w1JKsnhs1+1JGmQUR2RnAdcVVWrgau67YdIsgi4EFgHrAFOT7Kmr/8I4CTgP+ekYknSQKMKkvXAJd3zS4ANA8YcD+ysqlur6kHgsm7ePu8G3gTUbBYqSZraqILkyVW1G6D7+aQBY5YDt/Vt7+raSHIq8M2qumF/CyXZlGQ8yfiePXvaK5ckPcTi2dpxks8ATxnQ9dZhdzGgrZI8ttvHS4fZSVVtBjYDjI2NefQiSTNs1oKkql4yWV+SO5Isq6rdSZYBdw4Ytgs4om97BXA7cDSwCrghyb72LyU5vqq+NWMvQJI0lFGd2toCbOyebwQuHzDmOmB1klVJlgCnAVuq6itV9aSqWllVK+kFznGGiCSNxqiC5J3ASUluoffJq3cCJHlqkq0AVbUXOAe4AtgBfLyqto+oXknSJGbt1NZUqurbwIsHtN8OnNy3vRXYup99rZzp+iRJw/Ob7ZKkJgaJJKmJQSJJamKQSJKaGCSSpCYGiSSpiUEiSWpikEiSmhgkkqQmBokkqYlBIklqYpBIkpoYJJKkJgaJJKmJQSJJamKQSJKaGCSSpCYGiSSpiUEiSWpikEiSmhgkkqQmBokkqYlBIklqYpBIkpqkqkZdw5xJsgf4HnD3NKYfDtw1sxVpCocwvd/TgexAfU2jqmu2153p/c/U/lr3M935Le9hT6uqpZN1LqggAUiyuao2TWPeeFWNzUZNerjp/p4OZAfqaxpVXbO97kzvf6b217qfA/E9bCGe2vrbURegoTwaf08H6msaVV2zve5M73+m9te6nwPu36MFd0QyXR6RSJrPPCI5MGwedQGS1GDW3sM8IpEkNfGIRJLUxCCRJDVZMEGS5OIkdyb5txna38Ykt3SPjX3tq5J8sWv/WJIlM7GepIVtDt/DzkmyM0klOXyYfS2YIAE+DKx9pJOSfC7JygltTwTeDjwPOB54e5LDuu53Ae+uqtXAd4Gzpl+yJP3Ah5mb97B/Bl4CfGPYNRZMkFTV1cB3+tuSHJ3k00m2JbkmybOG3N3PAVdW1Xeq6rvAlcDaJAFeBPx1N+4SYMMMvQRJC9hcvId163y5qr7+SGpb/EgGPwptBs6uqluSPA/4M3pBsD/Lgdv6tnd1bT8KfK+q9k5ol6TZMNPvYdOyYIMkycHA84G/6h1IAPDDXd+ZwOu6tqcDW5M8CHytql4GhIerKdolaUbN0nvYtCzYIKF3Wu97VXXMxI6q+hDwIeidXwReNeFQbxfwwr7tFcDn6N0Q7dAki7ujkhXA7bNQuyTNxnvYtAtZkKrqHuBrSX4FID3PGXL6FcBLkxzWXaB6KXBF9b7d+Vngl7txG4HLZ7h0SZqV97Dp1rJggiTJR4F/AZ6ZZFeSs4AzgLOS3ABsB9YPs6+q+g7w+8B13eP3ujaANwOvT7KT3jWTD87sK5G0EM3Ve1iS1ybZRe8o5cYkH9hvbd4iRZLUYsEckUiSZodBIklqYpBIkpoYJJKkJgaJJKmJQaIFL8l9c7zeF2ZoPy9McneSLye5KckfDzFnQ5I1M7G+tI9BIs2wJFPeMaKqnj+Dy11TVccCxwKnJDlxP+M3AAaJZtRCvkWKNKkkRwMXAkuB+4FXV9VNSX4BeBuwBPg2cEZV3ZHkHcBTgZXAXUm+ChwJHNX9fE9VXdDt+76qOjjJC4F30Lu1zrOBbcDLq6qSnAyc3/V9CTiqqk6ZrN6qeiDJ9XQ33kvyamBTV+dO4BXAMcCpwM8keRvwS930h73Ohn90WoA8IpEG2wz8VlU9F3gDvbuqAnweOKE7CrgMeFPfnOcC66vq17vtZ9G7Xfe+v/dw0IB1jgXOpXeUcBRwYpLHAO8H1lXVC+i9yU+pu83FauDqrukTVfUTVfUcYAdwVlV9AdgCvLGqjqmq/5jidUpD84hEmmCqu6rSu23Ex5Iso/d/+1/rm7qlqh7o2/5UVX0f+H6SO4En07tZXr9/rapd3brX0zuiuQ+4tar27fuj9I4uBvmpJDcCzwTeWVXf6tqfneQPgEOBgxlwH6X9vE5paAaJ9HCT3lUVeC9wflVt6Ts1tc9/TRj7/b7n/8Pg/94GjRl0i+/JXFNVpyR5BvD5JJ+squvp/TW9DVV1Q5JX8dA7ve4z1euUhuapLWmC/dxV9RDgm93zjYPmz4CbgKP6/jzqr+1vQlV9FfhDejcNBXg8sLs7nXZG39B7u77Wu8dKP2CQSPDY7m6q+x6vZ/K7qr6D3qmga+hdCJ9x3emx1wCfTvJ54A7g7iGmXgT8dJJVwO8CX6T3J1T7L55fBryx+8jw0Uzz7rFSP+/+Kx2AkhxcVfeld/HiQuCWqnr3qOuSBvGIRDowvbq7+L6d3um094+4HmlSHpFIkpp4RCJJamKQSJKaGCSSpCYGiSSpiUEiSWryfyUA+LYAU/gfAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "graph=learn.recorder.plot(return_fig=True)\n",
    "graph.savefig('loss.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "learn.fit_one_cycle(5, 1e-2, moms=(0.8,0.7))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "learn.save('b1', return_path=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "learn.fit_one_cycle(5, 5e-3, moms=(0.8,0.7))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "learn.save('b2', return_path=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "learn.save_encoder('b2_enc', return_path=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (genomic_ulmfit)",
   "language": "python",
   "name": "genomic_ulmfit"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
